src/esperanza/finalizationstate.cpp

```diff
@@ -147,10 +149,10 @@ void FinalizationState::InstaJustify() {
   m_last_justified_epoch = m_current_epoch - 1;
 
   if (m_current_epoch > 1) {
-    uint32_t expected_finalized = m_current_epoch - 2;
-    if (GetCheckpoint(expected_finalized).m_is_justified) {
-      GetCheckpoint(expected_finalized).m_is_finalized = true;
-      m_last_finalized_epoch = expected_finalized;
+    uint32_t prev_justified = m_current_epoch - 2;
+    if (GetCheckpoint(prev_justified).m_is_justified) {
+      cp.m_is_finalized = true;
```

Comment:
@frolosofsky I think it's OK to keep previous epoch only justified. Actually, it's nice that we can look back and see if we reached finalization at that point or not. Otherwise, after we reached finalization we should go throw all checkpoints until the current one and mark them finalized but I don't think it's needed.

Reply:
Should't `GetCheckpoint(prev_justified)` become finalized as well?

Reply:
According to the change proposed here: https://github.com/dtr-org/unit-e/pull/1083/files#diff-924113d2ba8b7131c0af6bb2fe3c1d75R1101 it's finalized, but I suggest keepштп data consistent as much as possible.

Reply:
> @frolosofsky I think it's OK to keep previous epoch only justified. It's nice that we can look back and see if we reached finalization at that point or not. Otherwise, after we reached finalization, we should go through all checkpoints until the current one and mark them finalized, but I don't think it's needed.

I don't think so, honestly. You've already rewritten comments about justified->finalized because they were wrong, and why not change that in the code? Having #1083 merged will strictly limit amount of checkpoints.

(nit and up to you, I understand that this code works)

Reply:
The comment wasn't wrong but I changed as it seemed confusing. 

Eventually, I'll delete all checkpoints, but since it's harder to do, we have them. My point is that every FinalizationState (or checkpoints while we still have them) contains the information which was valid/up to date at the time this state was created/(considered the current one), and we shouldn't override old data (previous states) when we started to create new ones. Currently, this is true (I spot only 1 case when this rule is violated in FinalizationState but will fix it separately). 

I believe that relying on this immutability in this particular case will give us a safer code as it easier to maintain/less risky to unsynchronize. For instance, let's say we made old checkpoints `is_finalized=true` but then if we retrieve `FinalizationState` for these checkpoints, they won't have `is_finalized=true` flag.